drivers: pwm: add functions for capturing pwm pulse width and period

henrikbrixandersen opened the pull request:
Extend the PWM API with optional functions for capturing the PWM pulse width and period time.

Fixes #26026.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
drivers: pwm: add API for capturing pwm pulse width and period

Extend the PWM API with optional API functions for capturing PWM pulse
width and period cycles.

Fixes #26026.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

pabigot reviewed (commented):
Suggestion to make the API extension conditional on a Kconfig enabling it.  Example is in https://github.com/zephyrproject-rtos/zephyr/pull/25851/commits/48dc2fd11c1042c5a8d17bd1c715b5dd0cd7022d ("drivers: flash: add API to access SFDP region of serial flash devices") of #25851 where `CONFIG_FLASH_JESD216_API` enables the special case.

There're actually design considerations on how that should be implemented.  In that approach the API functions are only defined if the feature is enabled.  An alternative would be to always define them, but have them return `-ENOTSUP` if the feature is not enabled (that actually makes more sense at the moment).

henrikbrixandersen commented:
> Suggestion to make the API extension conditional on a Kconfig enabling it. Example is in [48dc2fd](https://github.com/zephyrproject-rtos/zephyr/commit/48dc2fd11c1042c5a8d17bd1c715b5dd0cd7022d) ("drivers: flash: add API to access SFDP region of serial flash devices") of #25851 where `CONFIG_FLASH_JESD216_API` enables the special case.

Thank you. I have added a Kconfig option for enabling this. Example implementation still pending.

pabigot commented on include/drivers/pwm.h:
```diff
@@ -43,6 +43,31 @@ typedef int (*pwm_pin_set_t)(struct device *dev, uint32_t pwm,
 			     pwm_flags_t flags);
 
+#ifdef CONFIG_PWM_CAPTURE_API
+
```
Per current policy these should not be excluded if the Kconfig isn't enabled.

pabigot commented on include/drivers/pwm.h:
```diff
@@ -87,6 +116,94 @@ static inline int z_impl_pwm_pin_set_cycles(struct device *dev, uint32_t pwm,
+						pwm_flags_t flags,
+						k_timeout_t timeout)
+{
+	struct pwm_driver_api *api;
```
The public API also should not be excluded.

I'm not sure if it's sufficient to leave the `__syscall()` declaration exposed but exclude the `z_impl` inline, or if the `z_impl` inline should be present but just return `-ENOTSUP`.   I've generally assumed the latter, for which I like the following as an idiom, but won't insist on it.
```
static inline int z_impl_foo (args)
{
   int ret = -ENOTSUP;

#ifdef CONFIG_SELECTOR
   blah blah
   ret = api->fn(args);
#endif

   return ret;
}
```

pabigot reviewed (commented):
I see this is still draft, but since I was looking at it here're the initial comments.  Will check again when there's an implementation.

pabigot commented on include/drivers/pwm.h:
```diff
@@ -55,11 +76,15 @@ typedef int (*pwm_get_cycles_per_sec_t)(const struct device *dev,
 
 /**
- * @brief Set the period and pulse width for a single PWM output.
+ * @brief Set the period and pulse width for a single PWM input.
```
Is this supposed to have changed from output to input?  The rest of the description seems to suggest it's an output.

pabigot reviewed (commented):
It doesn't seem quite right to have both a synchronous and an asynchronous API implementation.    Seems like that would produce duplication.  How about just the async as the implemented API, and have the sync wrapper use a stack signal structure and `k_poll()`?

coderkalyan commented:
@henrikbrixandersen What is the state of this? Does the API need any more work? Or is the wait because implementations haven't been finished yet?

henrikbrixandersen commented:
> @henrikbrixandersen What is the state of this? Does the API need any more work? Or is the wait because implementations haven't been finished yet?

I have had to rework the proposed API to meet real-life application needs (support for continuous capture and filtering through a callback mechanism). I have a couple of implementations almost ready. I will update this PR once the final issues have been worked out.

henrikbrixandersen committed:
drivers: pwm: mcux_ftm: add PWM capure support

Add PWM capture support to the NXP MCUX FlexTimer driver.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
tests: drivers: pwm: add PWM loopback test

Add test cases for the PWM capture API using PWM signal loopback.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
boards: arm: frdm_k64f: enable FlexTimer 0 as PWM

Enable FlexTimer 0 (FTM0) as PWM and setup PTC1 as FTM0 channel 0 for
use in PWM loopback test.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
CODEOWNERS: take ownership of PWM capture helper functions

Take ownership of the PWM capture helper functions.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

henrikbrixandersen committed:
MAINTAINERS: add myself as collaborator on PWM

Add myself as collaborator on the PWM code.

Signed-off-by: Henrik Brix Andersen <hebad@vestas.com>

pabigot commented on include/drivers/pwm.h:
```diff
@@ -22,13 +23,45 @@
+ */
+
+/** PWM pin capture captures period. */
+#define PWM_CAPTURE_TYPE_PERIOD		(1 << 1)
```
Is bit 0 reserved for some reason?

henrikbrixandersen replied:
Yes, it is used for the polarity (`PWM_POLARITY_NORMAL`/`PWM_POLARITY_INVERTED`), which can be set from DTS.

pabigot replied:
Ah.  OK, a comment to that effect could help (or defining `PWM_CAPTURE_POLARITY_foo` as aliases, since the devicetree constant has a different prefix).

pabigot commented on include/drivers/pwm.h:
```diff
@@ -22,13 +23,45 @@
+/** @endcond */
+
+/* PWM pin capture captures a single period/pulse width. */
+#define PWM_CAPTURE_MODE_SINGLE		(0 << 3)
```
There's some precedent for consistent use of `FOO_SHIFT` and `FOO_MASK` which would make this:

    #define PWM_CAPTURE_TYPE_SHIFT 1
    #define PWM_CAPTURE_TYPE_MASK (3U << PWM_CAPTURE_TYPE_SHIFT)
    #define PWM_CAPTURE_MODE_SHIFT 3
    #define PWM_CAPTURE_MODE_MASK (1U << PWM_CAPTURE_MODE_SHIFT)
    #define PWM_CAPTURE_MODE_SINGLE (0U << PWM_CAPTURE_MODE_SHIFT)
    #define PWM_CAPTURE_MODE_CONTINUOUS (1U << PWM_CAPTURE_MODE_SHIFT)

This reduces the magic numbers and makes it easier to change the bitfield assignments.

henrikbrixandersen replied:
Thanks, implemented.

pabigot commented on include/drivers/pwm.h:
```diff
@@ -89,6 +177,138 @@ static inline int z_impl_pwm_pin_set_cycles(const struct device *dev,
+					    pwm_capture_callback_handler_t cb,
+					    void *user_data)
+{
+	struct pwm_driver_api *api;
```
Why not combine the initialization into the declaration?

henrikbrixandersen replied:
Adopted.

pabigot commented on include/drivers/pwm.h:
```diff
@@ -186,6 +406,176 @@ static inline int pwm_pin_set_nsec(const struct device *dev, uint32_t pwm,
+ * @retval -EIO IO error while capturing.
+ * @retval -ERANGE If result is too large.
+ */
+
```
unnecessary blank line?  several occurrences in file.

henrikbrixandersen replied:
Fixed, thanks.

pabigot commented on include/drivers/pwm.h:
```diff
@@ -186,6 +406,176 @@ static inline int pwm_pin_set_nsec(const struct device *dev, uint32_t pwm,
+ * @brief Capture a single PWM period/pulse width in nanoseconds for a single
+ *        PWM input.
+ *
+ * This API function wraps calls to @a pwm_pin_capture_cycles() and @a
```
Do we need both nsec and usec implementations?   With 64 bit values nsec should cover all cases; the caller can divide if appropriate.

Though AFAICT this API already provides both, so I guess we have to continue with the duplication.

henrikbrixandersen replied:
Yeah, I added it both as a convenience and due to the API already supporting `cycles`, `nsec`, and `usec` variants for PWM generation.

pabigot commented on drivers/pwm/pwm_mcux_ftm.c:
```diff
@@ -27,12 +32,27 @@ struct mcux_ftm_config {
+	pwm_capture_callback_handler_t callback;
+	void *user_data;
+	uint32_t first_edge_overflows;
+	bool pulse_capture : 1;
```
Why use bitfield syntax here?

henrikbrixandersen replied:
That was a leftover from an earlier iteration where I had more boolean values in the same struct. Fixed.

pabigot reviewed (commented):
I'd like to see the mask/shift defines cleaned up a little, but overall I don't have concerns about this, though I also don't use PWM at all and can't verify the implementation or whether the API supports other capture-capable PWM hardware.

MaureenHelm commented on drivers/pwm/Kconfig:
```diff
@@ -21,6 +21,12 @@ config PWM_SHELL
 	help
 	  Enable the PWM related shell commands.
 
+config PWM_CAPTURE_API
```
The `_API` suffix seems weird, can we just call this `PWM_CAPTURE`?

henrikbrixandersen replied:
I'd like to, but I used `_API` to match our other, conditional APIs (`FLASH_JESD216_API`, `UART_ASYNC_API`, `LEGACY_TIMEOUT_API`, `POSIX_API`, ...).

carlescufi replied:
Agree with @MaureenHelm's point, we should instead move towards removing the `_API` suffix wherever it is not strictly necessary.

henrikbrixandersen replied:
I have changed it to `CONFIG_PWM_CAPTURE` instead.

pabigot commented on drivers/pwm/pwm_capture.c:
```diff
@@ -0,0 +1,85 @@
+	struct z_pwm_pin_capture_cb_data data;
+	int err;
+
+	if (flags & PWM_CAPTURE_MODE_CONTINUOUS) {
```
This idiom is fragile: it works only because continuous is the non-zero value for the bitfield.  In that case the subsequent bitwise or of single has no effect.

I'd rather see these sorts of things as `(flags & PWM_CAPTURE_MODE_MASK) == PWM_CAPTURE_MODE_CONTINUOUS)` as that's very clear.

henrikbrixandersen replied:
Thanks, fixed.

MaureenHelm commented on drivers/pwm/pwm_mcux_ftm.c:
```diff
@@ -174,7 +174,7 @@ static int mcux_ftm_pin_configure_capture(const struct device *dev,
 	param = &data->capture[pair].param;
 
-	if (flags & PWM_CAPTURE_MODE_CONTINUOUS) {
+	if ((flags & PWM_CAPTURE_MODE_MASK) == PWM_CAPTURE_MODE_CONTINUOUS) {
```
Looks like this change is in the wrong commit `tests: drivers: pwm: add PWM loopback test`

henrikbrixandersen replied:
Fixed.

coderkalyan commented:
@henrikbrixandersen By any chance are you planning to write an implementation for STM32 devices? I would really like to use this API on an STM32 device to replace my current GPIO interrupt "stopwatch" implementation, but I don't know very much at all about the low level STM32 HALs in order to be able to implement this driver :frowning_face:

henrikbrixandersen commented:
> @henrikbrixandersen By any chance are you planning to write an implementation for STM32 devices? I would really like to use this API on an STM32 device to replace my current GPIO interrupt "stopwatch" implementation, but I don't know very much at all about the low level STM32 HALs in order to be able to implement this driver ☹️

@coderkalyan Sorry, no. I have no plans for adding PWM capture support on STM32. Others can submit PRs for that, though, once this PR (hopefully) goes in.

anangl commented on include/drivers/pwm.h:
```diff
@@ -43,6 +76,56 @@ typedef int (*pwm_pin_set_t)(const struct device *dev, uint32_t pwm,
+ * Note: The callback handler will be called in interrupt context.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param flags PWM capture flags
+ * @param period Pointer to the memory to store the captured PWM period width
+ *               (in clock cycles). HW specific.
+ * @param pulse Pointer to the memory to store the captured PWM pulse width (in
+ *              clock cycles). HW specific.
```
These entries need to be corrected. The callback function signature specifies `pwm` parameter, not `flags`, and the other two parameters are named `period_cycles` and `pulse_cycles`, respectively.

henrikbrixandersen replied:
Good catch. Fixed, thanks.

henrikbrixandersen commented on include/drivers/pwm.h:
```diff
@@ -89,6 +177,131 @@ static inline int z_impl_pwm_pin_set_cycles(const struct device *dev,
+					    pwm_capture_callback_handler_t cb,
+					    void *user_data)
+{
+	struct pwm_driver_api *api = (struct pwm_driver_api *)dev->api;
```
Added.

henrikbrixandersen commented on include/drivers/pwm.h:
```diff
@@ -89,6 +177,131 @@ static inline int z_impl_pwm_pin_set_cycles(const struct device *dev,
+{
+	struct pwm_driver_api *api = (struct pwm_driver_api *)dev->api;
+
+	if (!api->pin_configure_capture) {
+		return -ENOTSUP;
```
Yes. The rationale for this is that there may be several different PWM controllers present in the system, some with PWM capture support, some without.

henrikbrixandersen commented on include/drivers/pwm.h:
```diff
@@ -89,6 +177,131 @@ static inline int z_impl_pwm_pin_set_cycles(const struct device *dev,
 	return api->pin_set(dev, pwm, period, pulse, flags);
 }
 
+#ifdef CONFIG_PWM_CAPTURE
```
The `pwm_pin_configure_capture()` function _is_ included in the documentation since the documentation build sets `CONFIG_PWM_CAPTURE`.

Attempting to use PWM capture in an application without `CONFIG_PWM_CAPTURE` should result in a linker error (similar to what we discussed in an API meeting a while back regarding the ADC API after the introduction of b7ce84930794a41ebe2c735f1d8c54f94712c7ba).

anangl replied:
> The `pwm_pin_configure_capture()` function _is_ included in the documentation since the documentation build sets `CONFIG_PWM_CAPTURE`.

Right, I missed that.

> Attempting to use PWM capture in an application without `CONFIG_PWM_CAPTURE` should result in a linker error (similar to what we discussed in an API meeting a while back regarding the ADC API after the introduction of [b7ce849](https://github.com/zephyrproject-rtos/zephyr/commit/b7ce84930794a41ebe2c735f1d8c54f94712c7ba)).

I recall that discussion but I don't recall that there was a conclusion that such attempts should result in linker errors. This is not the case for `CONFIG_SPI_ASYNC` or `CONFIG_UART_ASYNC_API`, for example. And I'm actually not convinced that linker errors are the best way to signal such attempts. To ensure that the required option is enabled, one can use a corresponding dependency in Kconfig or a build assertion in the code that uses this API, or one can check the option with `IS_ENABLED()` before doing the actual call. And I can imagine a scenario, not very common but still, where such optional API could be called from some precompiled library capable of handling properly the `-ENOTSUP` error returned when the API is not available. The approach with linker errors enforces that the caller code must be compiled with the same setting of the option as the driver code, so one would need two versions of such precompiled library.
That's only my opinion, though. Others can see it differently, so I'm not going to try to block this PR for this reason.

henrikbrixandersen replied:
The consensus, as I recall, was that the `#ifdef` should be around the `z_impl_*` function, but that the syscall should not be `#ifdef`'ed out. This will produce a linker error if trying to use the syscall without the Kconfig enabled. Here, the function in question is not a syscall (only an implementation) thus the entire function should be `#ifdef`'ed.

I believe the scenario, you list, is highly theoretical and will run into other, similar problems in many other parts of the kernel code.

Thalley replied:
@henrikbrixandersen You requested an opinion from Slack: Not familiar with the code but my general opinion is that it's best to avoid `#ifdef`s in the header files (basically the only prober usage of that in header files, IMO, is to optimize structs). 

If my application uses some function which isn't available due to a mismatch in the config, I'd rather get that message during compilation/linking, than at runtime, i.e. have a `#ifdef` around the implementation, rather than having the function return `-EOPNOTSUPP` (please don't use `-ENOTSUP` for this as that has a different meaning). There are some cases where the function may want to return `-EOPNOTSUPP` instead of being guarded by a `#ifdef` though. 

So my 2 cents:
1) Don't guard the function declaration
2) Guard the function implementation
2.a) If not possible/applicable to guard the function implementation, return `-EOPNOTSUPP`.

One compromise could be to mention the reliance of a specific KConfig in the function documentation, if you want.

pabigot replied:
Regarding `-ENOTSUP` see #23727; it's wrong, but it's unlikely to change.  Also `EOPNOTSUP` appears to be Linux-specific, while `ENOTSUP` is POSIX; the two do have the same value, but I don't want to see documentation referencing a non-POSIX identifier in the `E*` namespace.

I too would rather get build/link errors to determine when something is missing.  But "consensus" in Zephyr depends on who was attending the telecon where it was discussed, and generally isn't recorded anywhere, so it's a weak concept.

pabigot replied:
Looking back at the top comment per the slack request:  Where an API structure lacks the API pointer the wrapper should return `-ENOTSUP`, if other drivers with the same API may support the feature.  Existing practice of unconditionally invoking the pointer is perhaps unsafe.

But where an entire suite of functionality in all drivers using a particular subsystem driver API is disabled, I would rather discover use of the unsupported API as a link time error, as is currently done in this PR by eliding the function definition.  This seems more robust in the face of somebody changing the Kconfig to remove the support: I would want the application to fail if it required the support, and erroneous build asserts intended to check for this case presence may be incorrect or become outdated.

I do agree that the description of the conditionally-supported function should include a reference to the required Kconfig.  Based on existing use in `kernel.h` and [elsewhere](https://github.com/zephyrproject-rtos/zephyr/blob/6297c667c591055e6c2f88d8b9a2b01a05c9bb94/include/sys/cbprintf.h#L53) something like this seems reasonable:

```
 * @note @option{CONFIG_PWM_CAPTURE} must be selected for this function to be available.
```

The concern with linking to an external library that has different expectations is valid, but I don't think it's safe in Zephyr anyway. AFAIK we don't define a stable ABI.

Thalley replied:
> Regarding `-ENOTSUP` see #23727; it's wrong, but it's unlikely to change. Also `EOPNOTSUP` appears to be Linux-specific, while `ENOTSUP` is POSIX; the two do have the same value, but I don't want to see documentation referencing a non-POSIX identifier in the `E*` namespace.
> 
> I too would rather get build/link errors to determine when something is missing. But "consensus" in Zephyr depends on who was attending the telecon where it was discussed, and generally isn't recorded anywhere, so it's a weak concept.

Thanks for that information! I was actually going by the definition of the error codes in `lib/libc/minimal/include/errno.h` in Zephyr:

```
#define ENOTSUP 35      /* Unsupported value */
#define EOPNOTSUPP 45      /* Operation not supported on socket */
```

I won't be commenting more about this in this threads (seems to be out of scope), but I can't find much information about whether or not `EOPNOTSUPP` is POSIX, except that most sources I find on Google seem to say that it is indeed POSIX and not Linux (see also https://pubs.opengroup.org/onlinepubs/9699919799/).

henrikbrixandersen replied:
> I do agree that the description of the conditionally-supported function should include a reference to the required Kconfig.

Thank you for the references. I have added a note to the conditional API functions about enabling the Kconfig for them to be available.

henrikbrixandersen replied:
As per todays API meeting, the current [API design guidelines for conditional APIs](https://docs.zephyrproject.org/latest/guides/design_guidelines.html#conditional-data-and-apis) states the following:

> Function declarations that are available only when the option is enabled should be provided unconditionally. Add a note in the description that the function is available only when the specified feature is enabled, referencing the required Kconfig symbol by name. (In cases where the function is used but not enabled a link-time error will be generated.)

The implementation in this PR is consistent with the above design guidelines.

henrikbrixandersen commented on drivers/pwm/Kconfig:
```diff
@@ -21,6 +21,12 @@ config PWM_SHELL
 	help
 	  Enable the PWM related shell commands.
 
+config PWM_CAPTURE
+	bool "Provide API for PWM capture"
```
I would prefer not to introduce that. I find it highly confusing as a user when a given API cannot be enabled in Kconfig due to a "hidden" setting like that. What do others think?

henrikbrixandersen replied:
It just occurred to me that introducing a `CONFIG_PWM_SUPPORT_CAPTURE` or similar Kconfig option will not work well for PWM since there are PWM capture-only devices out there (see e.g. the NXP Kinetis Pulse Width Timer (PWT) in #31165), for which the driver Kconfig option itself will depend on `CONFIG_PWM_CAPTURE`. This will introduce circular dependencies.

henrikbrixandersen replied:
The consensus on todays API meeting was to go without a `CONFIG_PWM_SUPPORT_CAPTURE` Kconfig option, at least for now.

MaureenHelm commented:
> > @henrikbrixandersen By any chance are you planning to write an implementation for STM32 devices? I would really like to use this API on an STM32 device to replace my current GPIO interrupt "stopwatch" implementation, but I don't know very much at all about the low level STM32 HALs in order to be able to implement this driver ☹️
> 
> @coderkalyan Sorry, no. I have no plans for adding PWM capture support on STM32. Others can submit PRs for that, though, once this PR (hopefully) goes in.

@erwango can you take a look?

erwango commented:
> > > @henrikbrixandersen By any chance are you planning to write an implementation for STM32 devices? I would really like to use this API on an STM32 device to replace my current GPIO interrupt "stopwatch" implementation, but I don't know very much at all about the low level STM32 HALs in order to be able to implement this driver frowning_face
> > 
> > 
> > @coderkalyan Sorry, no. I have no plans for adding PWM capture support on STM32. Others can submit PRs for that, though, once this PR (hopefully) goes in.
> 
> @erwango can you take a look?

No plans on our side for now, @gmarull maybe?

henrikbrixandersen commented:
To demonstrate the PWM capture API on more hardware, I have added another driver implementation in #31165. I will not add it here since it currently has a few dependencies on other PRs to be merged.

coderkalyan commented:
Ah, I guess I'll have to learn the LL framework and implement the driver myself :cry: :joy: In the meantime, if anyone is kind enough to do it, it would be very much appreciated by me!

henrikbrixandersen commented:
@MaureenHelm Could you please re-approve? @zephyrbot cancelled your approval 👎

henrikbrixandersen commented:
@anangl Please re-review.
